### ISSUE TITLE
FR: [Auth] Allow to sign in with a token obtained from another App

### ISSUE THREAD

**Issue by jsayol** [firebase/firebase-js-sdk#49]
### [REQUIRED] Describe your environment

  * Operating System version: Linux
  * Firebase SDK version: 4.1.2
  * Firebase Product: auth

### [REQUIRED] Describe the problem
In certain scenarios, it could be useful to authenticate a `firebase.app.App` with a token obtained via `firebase.User#getToken()` from another App, when both Apps are initialized against the same Firebase project.

My use case for this is as follows: when a certain event happens, I want to close the connection to the database to stop receiving any new updates from whatever listeners I have attached, but I still need to listen to a particular location in the database that will alert me when I need to connect again.

At the moment there's no way to authenticate a second App with the same token/credentials already present in the current App. The only way to achieve this right now would be to either ask the user to reauthenticate when initializing the second App, or to initialize and authenticate both Apps at the same time when the user logs in.

_Related firebase-talk discussion: https://groups.google.com/forum/#!msg/firebase-talk/X5ry98EfK-w/8TnY2i8rCgAJ_

#### Steps to reproduce:
n/a

#### Relevant Code:
A possible solution would be to introduce a new `firebase.auth.Auth#signInWithIdToken()` method which would allow to do the following:
```ts
const options = {/* ... */};
const app1 = firebase.initializeApp(options, 'app1');
const app2 = firebase.initializeApp(options, 'app2');

// Let's assume "app1" is already authenticated
app1.auth().currentUser.getIdToken().then(token => app2.auth().signInWithIdToken(token));
```

Alternatively, a new `IdTokenAuthProvider` could be introduced, allowing to authenticate with the already existing `firebase.auth.Auth#signInWithCredential()`:
```ts
const options = {/* ... */};
const app1 = firebase.initializeApp(options, 'app1');
const app2 = firebase.initializeApp(options, 'app2');

// Let's assume "app1" is already authenticated
app1.auth().currentUser.getIdToken().then(token => {
  const cred = firebase.auth.IdTokenAuthProvider.credential(token);
  app2.auth().signInWithCredential(cred);
});
```

Introducing a new `AuthProvider` for this might be confusing for the end-user though, so my preferred solution would be `signInWithIdToken`.

**Comment by bojeil-google** [comment]
I am sure this will be convenient, but it has security implications and will open the door for abuse. We have already thought about this. It's a common feature request. Having the ability to use a short lived token to generate a permanent session without the ability to revoke it (there is no way to revoke a Firebase ID tokens like OAuth providers do) is not something we are comfortable supporting or encouraging. 

Instead, we could support something like using postMessage from one window to another to transfer an auth state securely while checking origin (that will work for the same browser though) and could work with chrome extensions too. We could also use cross subdomain cookies, etc.

For other cases, you can always use custom tokens and build your own endpoints to support this as long as you are comfortable with the risks. You can create some endpoint that takes an ID token and verifies it and returns a custom token to transfer that state. I do not want to recommend it though.

**Comment by mbleigh** [comment]
Just spitballing here, what about something like `signInWithApp(app:
FirebaseApp)` so that you can pass a (different) FirebaseApp instance and
it will reuse the auth on that app?

**Comment by bojeil-google** [comment]
So the use case here is to transfer auth state from one app to another in the same window?

**Comment by mbleigh** [comment]
Yep.

**Comment by bojeil-google** [comment]
I suppose that would be doable. I don't see any issues with it.

**Comment by jsayol** [comment]
Yes, the use case I had in mind here would use different apps in the same window. Sorry if that wasn't clear.

Just to illustrate what I'd like to achieve, it is technically already possible to do this albeit in a very hacky way (_note to anyone else reading: do not do this_):

```ts
const options = {
  apiKey: '...',
  /* ... */
};

const app1 = firebase.initializeApp(options, 'app1');

window.localStorage.setItem(
  `firebase:authUser:${options.apiKey}:app2`,
  window.localStorage.getItem(`firebase:authUser:${options.apiKey}:app1`)
);

const app2 = firebase.initializeApp(options, 'app2');
```

So basically I'd just like a clean and supported way to achieve this.

I liked Michael's approach. Instead of messing with tokens like I initially suggested, there could be a method to simply copy the auth state from one app to another and let the SDK internally handle how that's done.

**Comment by ArtemBogush** [comment]
Will this allow to get a user from let's say chrome extension and to send it to the website to automatically login there?

**Comment by bojeil-google** [comment]
No it won't. This is meant only for the same window. We are looking into a solution to the other case you mentioned.
